pkg/storegateway/indexheader/threadpool.go

```diff
@@ -0,0 +1,71 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+// Provenance-includes-location: https://github.com/cpg1111/threadpool-go/blob/master/threadpool.go
+// Provenance-includes-license: MIT
+// Provenance-includes-copyright: Christian Grabowski
+
+package indexheader
+
+import (
+	"context"
+	"errors"
+	"runtime"
+)
+
+type Threadpool struct {
+	ctx        context.Context
+	cancel     context.CancelFunc
+	pool       chan *OSThread
+	numThreads int
+}
+
+func NewThreadPool(ctx context.Context, cancel context.CancelFunc, num int) (*Threadpool, error) {
+	if num <= 0 {
+		return nil, nil
+	}
+
+	if num >= runtime.GOMAXPROCS(0) {
+		return nil, errors.New("threadpool size must be GOMAXPROCS - 1 at most")
+	}
+	if cancel == nil {
+		ctx, cancel = context.WithCancel(ctx)
+	}
+
+	tp := &Threadpool{
+		ctx:        ctx,
+		cancel:     cancel,
+		pool:       make(chan *OSThread, num),
+		numThreads: num,
+	}
+
+	for i := 0; i < num; i++ {
+		childCtx, childCancel := context.WithCancel(ctx)
+		t := NewOSThread(childCtx, childCancel)
+		t.Start()
+		tp.pool <- t
+	}
+
+	return tp, nil
+}
+
+func (t *Threadpool) start() {
+	for range t.ctx.Done() {
+		for i := 0; i < t.numThreads; i++ {
+			thread := <-t.pool
+			thread.Join()
+		}
+	}
+}
```

Comment:
Can you explain what this does?

As far as I understand, if you call start on a fresh Threadpool:
It pulls each thread from the pool, and calls `Join` on it, which is:
```go
func (o *OSThread) Join() {
	for range o.ctx.Done() {
		tasks := o.localTasks.Load()
		if tasks == 0 {
			break
		}
	}

	o.cancel()
}
```

Since there are no tasks, it will break on the first iteration of the loop, and cancel the local context and return.

Once this is done, the thread is stopped (because context is canceled) and nobody will put more threads in the pool (the only way of putting threads is in New() or by returning a thread from a call, but we have _joined_ that one on the first iteration)

Reply:
`ctx.Done()` is a channel so the loop body is blocked and doesn't run until the context is cancelled.

Reply:
The use of child context here is confusing though. I'm going to work on making this a bit more straightforward.

Reply:
Thank you. And yes, I misinterpreted the loop on ctx.Done() :facepalm: